File: src/hf_fetcher.py
"""Fetch public Hugging Face Hub data for a given username."""

import logging
from dataclasses import dataclass, field
from typing import List, Optional

from huggingface_hub import HfApi
from huggingface_hub.utils import HfHubHTTPError

logger = logging.getLogger(__name__)

MAX_ITEMS = 200


@dataclass
class UserProfile:
    username: str
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    num_followers: int = 0
    num_following: int = 0
    num_discussions: int = 0
    is_pro: bool = False


@dataclass
class RepoItem:
    id: str
    likes: int = 0
    downloads: int = 0
    tags: List[str] = field(default_factory=list)
    description: Optional[str] = None
    pipeline_tag: Optional[str] = None


@dataclass
class HfProfileData:
    user: UserProfile
    models: List[RepoItem]
    datasets: List[RepoItem]
    spaces: List[RepoItem]

    @property
    def total_model_likes(self) -> int:
        return sum(m.likes for m in self.models)

    @property
    def total_model_downloads(self) -> int:
        return sum(m.downloads for m in self.models)

    @property
    def total_space_likes(self) -> int:
        return sum(s.likes for s in self.spaces)

    @property
    def total_likes(self) -> int:
        return self.total_model_likes + self.total_space_likes + sum(d.likes for d in self.datasets)

    @property
    def total_downloads(self) -> int:
        return self.total_model_downloads + sum(d.downloads for d in self.datasets)


def _get_user_profile(api: HfApi, username: str) -> UserProfile:
    try:
        overview = api.get_user_overview(username)
    except HfHubHTTPError as e:
        if e.response.status_code == 404:
            raise ValueError(f"User '{username}' not found on Hugging Face.") from e
        raise
    except Exception as e:
        logger.warning("Failed to fetch user overview for %s: %s", username, e)
        raise ValueError(f"Could not fetch profile for '{username}'.") from e

    return UserProfile(
        username=overview.username,
        display_name=overview.fullname or overview.username,
        avatar_url=overview.avatar_url,
        num_followers=overview.num_followers or 0,
        num_following=overview.num_following or 0,
        num_discussions=overview.num_discussions or 0,
        is_pro=overview.is_pro or False,
    )


def _repo_item_from_model(item) -> RepoItem:
    return RepoItem(
        id=getattr(item, "id", ""),
        likes=getattr(item, "likes", 0) or 0,
        downloads=getattr(item, "downloads", 0) or 0,
        tags=list(getattr(item, "tags", []) or []),
        description=getattr(item, "description", None) or None,
        pipeline_tag=getattr(item, "pipeline_tag", None) or None,
    )


def _repo_item_from_dataset(item) -> RepoItem:
    return RepoItem(
        id=getattr(item, "id", ""),
        likes=getattr(item, "likes", 0) or 0,
        downloads=getattr(item, "downloads", 0) or 0,
        tags=list(getattr(item, "tags", []) or []),
        description=getattr(item, "description", None) or None,
    )


def _repo_item_from_space(item) -> RepoItem:
    return RepoItem(
        id=getattr(item, "id", ""),
        likes=getattr(item, "likes", 0) or 0,
        downloads=0,
        tags=list(getattr(item, "tags", []) or []),
        description=getattr(item, "description", None) or None,
    )


def fetch_hf_profile(username: str) -> HfProfileData:
    username = username.strip().lstrip("@")
    if not username:
        raise ValueError("Username is required.")

    api = HfApi()
    user = _get_user_profile(api, username)

    try:
        models = [
            _repo_item_from_model(m)
            for m in api.list_models(author=username, limit=MAX_ITEMS)
        ]
    except Exception as e:
        logger.warning("Failed to fetch models for %s: %s", username, e)
        models = []

    try:
        datasets = [
            _repo_item_from_dataset(d)
            for d in api.list_datasets(author=username, limit=MAX_ITEMS)
        ]
    except Exception as e:
        logger.warning("Failed to fetch datasets for %s: %s", username, e)
        datasets = []

    try:
        spaces = [
            _repo_item_from_space(s)
            for s in api.list_spaces(author=username, limit=MAX_ITEMS)
        ]
    except Exception as e:
        logger.warning("Failed to fetch spaces for %s: %s", username, e)
        spaces = []

    return HfProfileData(user=user, models=models, datasets=datasets, spaces=spaces)

File: web/src/lib/hf-fetcher.ts
const MAX_ITEMS = 200;

export type UserProfile = {
  username: string;
  displayName: string | null;
  avatarUrl: string | null;
  numFollowers: number;
  numFollowing: number;
  numDiscussions: number;
  isPro: boolean;
};

export type RepoItem = {
  id: string;
  likes: number;
  downloads: number;
  tags: string[];
  description: string | null;
  pipelineTag: string | null;
};

export type HfProfileData = {
  user: UserProfile;
  models: RepoItem[];
  datasets: RepoItem[];
  spaces: RepoItem[];
  totalModelLikes: number;
  totalModelDownloads: number;
  totalSpaceLikes: number;
  totalLikes: number;
  totalDownloads: number;
};

async function fetchJson<T>(url: string): Promise<T> {
  const res = await fetch(url, {
    headers: { Accept: "application/json" },
    next: { revalidate: 300 },
  });
  if (!res.ok) {
    if (res.status === 404) throw new Error(`User not found on Hugging Face.`);
    throw new Error(`Hugging Face API error (${res.status})`);
  }
  return res.json() as Promise<T>;
}

type OverviewResponse = {
  user: string;
  fullname?: string;
  avatarUrl?: string;
  numFollowers?: number;
  numFollowing?: number;
  isPro?: boolean;
};

type ModelItem = {
  id: string;
  likes?: number;
  downloads?: number;
  tags?: string[];
  description?: string;
  pipeline_tag?: string;
};

type DatasetItem = {
  id: string;
  likes?: number;
  downloads?: number;
  tags?: string[];
  description?: string;
};

type SpaceItem = {
  id: string;
  likes?: number;
  tags?: string[];
  description?: string;
};

export async function fetchHfProfile(username: string): Promise<HfProfileData> {
  const clean = username.trim().replace(/^@/, "");
  if (!clean) throw new Error("Username is required.");

  let overview: OverviewResponse;
  try {
    overview = await fetchJson<OverviewResponse>(
      `https://huggingface.co/api/users/${encodeURIComponent(clean)}/overview`,
    );
  } catch {
    throw new Error(`User '${clean}' not found on Hugging Face.`);
  }

  const [modelsRaw, datasetsRaw, spacesRaw] = await Promise.all([
    fetchJson<ModelItem[]>(
      `https://huggingface.co/api/models?author=${encodeURIComponent(clean)}&limit=${MAX_ITEMS}`,
    ).catch(() => [] as ModelItem[]),
    fetchJson<DatasetItem[]>(
      `https://huggingface.co/api/datasets?author=${encodeURIComponent(clean)}&limit=${MAX_ITEMS}`,
    ).catch(() => [] as DatasetItem[]),
    fetchJson<SpaceItem[]>(
      `https://huggingface.co/api/spaces?author=${encodeURIComponent(clean)}&limit=${MAX_ITEMS}`,
    ).catch(() => [] as SpaceItem[]),
  ]);

  const models: RepoItem[] = modelsRaw.map((m) => ({
    id: m.id,
    likes: m.likes ?? 0,
    downloads: m.downloads ?? 0,
    tags: m.tags ?? [],
    description: m.description ?? null,
    pipelineTag: m.pipeline_tag ?? null,
  }));

  const datasets: RepoItem[] = datasetsRaw.map((d) => ({
    id: d.id,
    likes: d.likes ?? 0,
    downloads: d.downloads ?? 0,
    tags: d.tags ?? [],
    description: d.description ?? null,
    pipelineTag: null,
  }));

  const spaces: RepoItem[] = spacesRaw.map((s) => ({
    id: s.id,
    likes: s.likes ?? 0,
    downloads: 0,
    tags: s.tags ?? [],
    description: s.description ?? null,
    pipelineTag: null,
  }));

  const totalModelLikes = models.reduce((s, m) => s + m.likes, 0);
  const totalModelDownloads = models.reduce((s, m) => s + m.downloads, 0);
  const totalSpaceLikes = spaces.reduce((s, sp) => s + sp.likes, 0);
  const totalLikes =
    totalModelLikes + totalSpaceLikes + datasets.reduce((s, d) => s + d.likes, 0);
  const totalDownloads =
    totalModelDownloads + datasets.reduce((s, d) => s + d.downloads, 0);

  return {
    user: {
      username: overview.user || clean,
      displayName: overview.fullname ?? overview.user ?? clean,
      avatarUrl: overview.avatarUrl ?? null,
      numFollowers: overview.numFollowers ?? 0,
      numFollowing: overview.numFollowing ?? 0,
      numDiscussions: 0,
      isPro: overview.isPro ?? false,
    },
    models,
    datasets,
    spaces,
    totalModelLikes,
    totalModelDownloads,
    totalSpaceLikes,
    totalLikes,
    totalDownloads,
  };
}
